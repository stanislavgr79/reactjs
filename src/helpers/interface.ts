export interface IMovie {
  [x: string]: any;
  id?: number;
  title: string;
  tagline?: string;
  vote_average?: number;
  vote_count?: number;
  release_date: string;
  movieUrl: string;
  genres: string[] | { label: string, value: string }[];
  poster_path?: string | undefined;
  overview: string;
  runtime: number | string | undefined;
}
