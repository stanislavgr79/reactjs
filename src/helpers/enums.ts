/* eslint-disable no-unused-vars */

/**
 * Enum for common SelectGenres.
 * @readonly
 * @enum {{label: string, value: string}}
 */
export const selectGenres = Object.freeze([
  { label: 'Action', value: 'Action' },
  { label: 'Adventure', value: 'Adventure' },
  { label: 'Animation', value: 'Animation' },
  { label: 'Comedy', value: 'Comedy' },
  { label: 'Documentary', value: 'Documentary' },
  { label: 'Drama', value: 'Drama' },
  { label: 'Family', value: 'Family' },
  { label: 'Horror', value: 'Horror' },
  { label: 'Mystery', value: 'Mystery' },
  { label: 'Thriller', value: 'Thriller' },
]);

/**
 * Enum for common Genres.
 * @readonly
 * @enum { string }
 */
export enum Genres {
  ALL = '',
  DOCUMENTARY = 'documentary',
  COMEDY = 'comedy',
  HORROR = 'horror',
  CRIME = 'crime',
}

/**
 * Enum for common SortBy.
 * @readonly
 * @enum { string }
 */
export enum SortBy {
  TITLE = 'title',
  RELEASE = 'release_date',
  RUNTIME = 'runtime',
  RATING = 'vote_average',
}

/**
 * Enum for common SortOrder.
 * @readonly
 * @enum { string }
 */
export enum SortOrder {
  ASC = 'asc',
  DESC = 'desc',
}
