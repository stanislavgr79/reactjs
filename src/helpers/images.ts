const images = [
  {
    name: 'Ricky and Morty.jpg',
    value: './resources/images/Ricky and Morty.jpg',
  },
  {
    name: 'Pulp Fiction.jpg',
    value: './resources/images/Pulp Fiction.jpg',
  },
  {
    name: 'Kill Bill.jpg',
    value: './resources/images/Kill Bill.jpg',
  },
  {
    name: 'Avengers.jpg',
    value: './resources/images/Avengers.jpg',
  },
  {
    name: 'Inception.jpg',
    value: './resources/images/Inception.jpg',
  },
  {
    name: 'Reservoir Dogs.jpg',
    value: './resources/images/Reservoir Dogs.jpg',
  },
];

export default images;
