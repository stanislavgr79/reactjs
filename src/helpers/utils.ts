export const createClassList = (classesArray: string[]) =>
  classesArray
    .filter((className) => typeof className === 'string' && className.length > 0)
    .join(' ')
    .trim();

export function transformGenresToStringArray(genres: { value: string, label: string }[]) {
  const genresArray: string[] = [];
  genres.map((it) => genresArray.push(it.value));
  return genresArray;
}

export function transformGenresToArrayObjects(genres: string[]) {
  const genresArray: { value: string, label: string }[] = [];
  genres.map((it: string) =>
    genresArray.push({
      label: it,
      value: it,
    }),
  );
  return genresArray;
}
