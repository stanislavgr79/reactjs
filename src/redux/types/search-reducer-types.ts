export const UPDATE_SEARCH_QUERY = 'UPDATE_SEARCH_QUERY';
export const UPDATE_LOCATION = 'UPDATE_LOCATION';

export interface SearchState {
  search: {
    value: string,
  };
  location: string;
}

interface UpdateSearchQuery {
  type: typeof UPDATE_SEARCH_QUERY;
  search: {
    value: string,
  };
}

interface UpdateLocation {
  type: typeof UPDATE_LOCATION;
  payload: string;
}

export type SearchActionTypes = UpdateSearchQuery | UpdateLocation;
