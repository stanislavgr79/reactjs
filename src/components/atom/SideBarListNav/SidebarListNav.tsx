import React, { useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Genres } from '../../../helpers/enums';
import { updateSelectedIndexListNav } from '../../../redux/actions/sidebar-actions';
import { AppState } from '../../../redux/redux-store';

import './SidebarListNav.scoped.scss';

interface IProps {
  defaultValue: Genres;
  updateCurrentGenre: (genre: Genres) => void;
}

export default function SidebarListNav({ defaultValue, updateCurrentGenre }: IProps): JSX.Element {
  const dispatch = useDispatch();
  const { selectedIndexListNav } = useSelector((store: AppState) => {
    return store.sidebar;
  });
  const sideGenre = Object.values(Genres);

  const handleChange = useCallback(
    (index: number) => {
      dispatch(updateSelectedIndexListNav(index));
    },
    [dispatch],
  );

  useEffect(() => {
    sideGenre[selectedIndexListNav] !== Genres.ALL
      ? updateCurrentGenre(sideGenre[selectedIndexListNav])
      : updateCurrentGenre(Genres.ALL);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedIndexListNav]);

  return (
    <div className="sidebar-list-nav">
      <form
        className="side-genre-list"
        onChange={(event) => event.preventDefault()}
        defaultValue={defaultValue}
      >
        {Object.entries(Genres).map(([key, value], index: number) => (
          <option
            className={selectedIndexListNav == index ? 'selected' : ''}
            id={`${value}-${index}`}
            value={value}
            key={`sidebar-genres-${index}`}
            onClick={() => handleChange(index)}
          >
            {key}
          </option>
        ))}
      </form>
    </div>
  );
}
