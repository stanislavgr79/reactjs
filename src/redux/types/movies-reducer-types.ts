import { IMovie } from '../../helpers/interface';

export const ADD_TO_MOVIES = 'ADD_TO_MOVIES';
export const DELETE_FROM_MOVIES = 'DELETE_FROM_MOVIES';
export const UPDATE_MOVIE_IN_MOVIES = 'UPDATE_MOVIE_IN_MOVIES';
export const UPDATE_STATUS_MOVIE_POPUP = 'UPDATE_STATUS_MOVIE_POPUP';
export const UPDATE_SELECTED_MOVIE_ID = 'UPDATE_SELECTED_MOVIE_ID';

export interface MoviesState {
  movies: IMovie[];
  moviesIsLoading: boolean;
  showPopup: boolean;
  selectedMovieId: number;
  movie: IMovie | undefined;
}

interface AddMovieToMovies {
  type: typeof ADD_TO_MOVIES;
  payload: IMovie;
}

interface DeleteFromMovies {
  type: typeof DELETE_FROM_MOVIES;
  payload: number;
}

interface UpdateMovieInMovies {
  type: typeof UPDATE_MOVIE_IN_MOVIES;
  payload: IMovie;
}

interface UpdateStatusMoviePopup {
  type: typeof UPDATE_STATUS_MOVIE_POPUP;
  payload: boolean;
}

interface UpdateSelectedMovieId {
  type: typeof UPDATE_SELECTED_MOVIE_ID;
  payload: number;
}

export type MoviesActionTypes =
  | AddMovieToMovies
  | DeleteFromMovies
  | UpdateMovieInMovies
  | UpdateStatusMoviePopup
  | UpdateSelectedMovieId;
