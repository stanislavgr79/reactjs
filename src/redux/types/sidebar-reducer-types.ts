import { Genres, SortBy, SortOrder } from '../../helpers/enums';

/* eslint-disable no-unused-vars */
export const UPDATE_SELECT_GENRE = 'UPDATE_SELECT_GENRE';
export const UPDATE_SORT_BY = 'UPDATE_SORT_BY';
export const UPDATE_SORT_ORDER = 'UPDATE_SORT_ORDER';
export const UPDATE_SELECTED_INDEX_LIST_NAV = 'UPDATE_SELECTED_INDEX_LIST_NAV';

export interface SidebarState {
  genre: Genres;
  sortBy: SortBy;
  sortOrder: SortOrder;
  selectedIndexListNav?: number;
}

interface UpdateCurrentGenre {
  type: typeof UPDATE_SELECT_GENRE;
  payload: Genres;
}

interface UpdateCurrentSortBy {
  type: typeof UPDATE_SORT_BY;
  payload: SortBy;
}

interface UpdateSortOrder {
  type: typeof UPDATE_SORT_ORDER;
  payload: SortOrder;
}

interface UpdateSelectedIndexListNav {
  type: typeof UPDATE_SELECTED_INDEX_LIST_NAV;
  payload: number;
}

export type SidebarActionTypes =
  | UpdateCurrentGenre
  | UpdateCurrentSortBy
  | UpdateSortOrder
  | UpdateSelectedIndexListNav;
