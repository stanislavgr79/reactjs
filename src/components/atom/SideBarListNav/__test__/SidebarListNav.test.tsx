import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import storeReal from '../../../../redux/redux-store';
// eslint-disable-next-line max-len
import { SidebarActionTypes, UPDATE_SELECTED_INDEX_LIST_NAV } from '../../../../redux/types/sidebar-reducer-types';

import { render, screen, RenderOptions } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import SidebarListNav from '../SidebarListNav';
import { Genres } from '../../../../helpers/enums';

describe('SidebarListNav', () => {
  const initialState = { sidebar: { selectedIndexListNav: 0 } };
  const mockStore = configureStore();
  let container: Element 
    | DocumentFragment 
    // eslint-disable-next-line prettier/prettier
    | RenderOptions<typeof import('@testing-library/dom/types/queries'), HTMLElement>
    | null;

  const updateCurrentGenreMock = jest.fn();
  
  afterEach(() => {
    container = null;
    jest.clearAllMocks();
  });

  test('render', async () => {
    const store = mockStore(initialState);

    container = render(
      <Provider store={store}>
          <SidebarListNav defaultValue={Genres.ALL} updateCurrentGenre={updateCurrentGenreMock} />
      </Provider>,
    );
    
    expect(container.baseElement).toMatchSnapshot();
  });

  test('render Genres.CRIME select', async () => {
    const store = storeReal;
    
    container = render(
      <Provider store={store}>
          <SidebarListNav defaultValue={Genres.ALL} updateCurrentGenre={updateCurrentGenreMock} />
      </Provider>,
    );

    userEvent.click(screen.getByText('CRIME'));
    expect(container.baseElement).toMatchSnapshot();
    expect(container.baseElement)
      .toContainHTML('<option class="selected" id="crime-4" value="crime">CRIME</option>');
  });

  test('handleChange function', async () => {
    const store = mockStore(initialState);

    jest.mock('../../../../redux/actions/sidebar-actions', () => ({
      updateSelectedIndexListNav: (selectedIndexListNav: number): SidebarActionTypes => {
        return {
          type: UPDATE_SELECTED_INDEX_LIST_NAV,
          payload: selectedIndexListNav,
        };
      },
    }));

    render(
      <Provider store={store}>
          <SidebarListNav defaultValue={Genres.ALL} updateCurrentGenre={updateCurrentGenreMock} />
      </Provider>,
    );
    
    userEvent.click(screen.getByText('CRIME'));
    const expectedActions = [{ type: 'UPDATE_SELECTED_INDEX_LIST_NAV', payload: 4 }];
    expect(store.getActions()).toEqual(expectedActions);
  });
});
