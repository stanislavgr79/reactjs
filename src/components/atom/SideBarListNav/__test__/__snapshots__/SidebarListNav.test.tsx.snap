// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarListNav render 1`] = `
<body>
  <div>
    <div
      class="sidebar-list-nav"
    >
      <form
        class="side-genre-list"
      >
        <option
          class="selected"
          id="-0"
          value=""
        >
          ALL
        </option>
        <option
          class=""
          id="documentary-1"
          value="documentary"
        >
          DOCUMENTARY
        </option>
        <option
          class=""
          id="comedy-2"
          value="comedy"
        >
          COMEDY
        </option>
        <option
          class=""
          id="horror-3"
          value="horror"
        >
          HORROR
        </option>
        <option
          class=""
          id="crime-4"
          value="crime"
        >
          CRIME
        </option>
      </form>
    </div>
  </div>
</body>
`;

exports[`SidebarListNav render Genres.CRIME select 1`] = `
<body>
  <div>
    <div
      class="sidebar-list-nav"
    >
      <form
        class="side-genre-list"
      >
        <option
          class=""
          id="-0"
          value=""
        >
          ALL
        </option>
        <option
          class=""
          id="documentary-1"
          value="documentary"
        >
          DOCUMENTARY
        </option>
        <option
          class=""
          id="comedy-2"
          value="comedy"
        >
          COMEDY
        </option>
        <option
          class=""
          id="horror-3"
          value="horror"
        >
          HORROR
        </option>
        <option
          class="selected"
          id="crime-4"
          value="crime"
        >
          CRIME
        </option>
      </form>
    </div>
  </div>
</body>
`;
